src/nlua_shiplog.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * See Licensing and Copyright notice in naev.h
+ */
+
+/**
+ * @file nlua_shiplog.c
+ *
+ * @brief Handles the shiplog Lua bindings.
+ */
+
+
+#include "nlua_shiplog.h"
+
+#include "naev.h"
+
+#include <stdlib.h>
+#include <stdio.h>
+#include "nstring.h"
+#include <math.h>
+
+#include <lua.h>
+#include <lauxlib.h>
+
+#include "nlua.h"
+#include "nlua_hook.h"
+#include "nlua_player.h"
+#include "nlua_tk.h"
+#include "nlua_faction.h"
+#include "nlua_system.h"
+#include "nlua_tex.h"
+#include "nlua_camera.h"
+#include "nlua_music.h"
+#include "nlua_bkg.h"
+#include "nlua_misn.h"
+#include "player.h"
+#include "mission.h"
+#include "log.h"
+#include "rng.h"
+#include "toolkit.h"
+#include "land.h"
+#include "nxml.h"
+#include "nluadef.h"
+#include "music.h"
+#include "gui_osd.h"
+#include "npc.h"
+#include "array.h"
+#include "ndata.h"
+#include "shiplog.h"
+
+int shiplog_loadShiplog( nlua_env env );
+static int shiplog_createLog( lua_State *L );
+static int shiplog_appendLog( lua_State *L );
+static const luaL_Reg shiplog_methods[] = {
+   { "createLog", shiplog_createLog },
+   { "appendLog", shiplog_appendLog },
+   {0,0}
+}; /**< Shiplog Lua methods. */
+
+
+/**
+ * @brief Registers the lua libraries.
+ *
+ *    @param L Lua state.
+ *    @return 0 on success.
+ */
+int shiplog_loadLibs( nlua_env env )
+{
+   nlua_loadStandard(env);
+   nlua_loadShiplog(env);
+   return 0;
+}
+/*
+ * individual library loading
+ */
+/**
+ * @brief Loads the mission Lua library.
+ *    @param L Lua state.
+ */
+int nlua_loadShiplog( nlua_env env )
+{
+   nlua_register(env, "shiplog", shiplog_methods, 0);
+   return 0;
+}
+
+
+/**
+ * @brief Creates a shiplog for this mission.
+ *
+ *    @luatparam string name Name for this log.
+ *    @luatparam string type Type of log (e.g travel, trade, etc, can be anything.)
+ *    @luatparam number overwrite Whether to overwrite existing mission with this logname and logtype.  Warning, removes previous entries of this logname and type if 1, or logs of this type if 2.
```

Comment:
The meaning of this is unclear; reading this I have no idea what this argument does. Also, you shouldn't use numbers that could be 1 or 2, it would be better to just have two separate booleans (which would also mean that `nil` would be valid).

🕵️

Reply:
ok, I'll split it into 2 bools.
Basically, if a user starts a mission and then aborts it, the logset will remain (which is what we want to happen).
If they then restart it at a later date, the existing log can then be replaced with the new attempt, if overwrite=1.  Or if overwrite=0, then the old logset will remain, and a new one will be created.

Overwrite=2 means that all logsets of this type will be replaced by the new logset.  This is more of an edge case, so less likely to be used.  In fact, I could probably remove it for now.

Reply:
> If they then restart it at a later date, the existing log can then be replaced with the new attempt, if overwrite=1.

Wait, what? I don't think we want this. Logs should *never* be automatically deleted.

> Overwrite=2 means that all logsets of this type will be replaced by the new logset.

Same here. I think you should remove these options altogether. Logs should be permanent.

🕵️

Reply:
Except if we ever implement logs for throwaway missions.  Might happen in the future, we just don't know.  I'll sort something out.  But for the non-throw-away missions, we can just set overwrite=0...

Is there a concept of flags in the lua?  i.e. in the same way as in c we would & flags.  If so, I could replace overwrite with flags, and then would allow for future expansion, as @PhoenixRiver mentions below.

Reply:
No, even if you did implement them for throwaway missions, they should be permanent.

Flags aren't something you should be doing in Lua. It's technically possible, but it would be a mess. You can't think of Lua like C, code is written for Lua completely differently. If you needed multiple conditions in one arg, the rest of Naev handles multiple cases or whatever by strings. But in a case like this, booleans would make the most sense.

Again though, I don't think logs should *ever* be removed. They should be permanent. If you're having to remove logs, that says that the logs are unimportant and shouldn't have been made in the first place.

🕵️